### ISSUE TITLE
TypeSet is always be reordered

### ISSUE THREAD

**Issue by merceskoba** [hashicorp/terraform#27018]
### Terraform Version

```tf
Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/google v3.48.0
```

### Terraform Configuration Files

I add host_rule for hosts "beta.domain.tld"
```tf
resource "google_compute_url_map" "url_map_a" {
  project = "playground-public"
  name = "url-map-a"
  default_service = google_compute_backend_service.playground_a.id

  host_rule {
    hosts = ["sdwr.domain.tld"]
    path_matcher = "spaths"
  }

  host_rule {
    hosts = ["beta.domain.tld"]
    path_matcher = "bpaths"
  }

  path_matcher {
    name = "spaths"
    default_service = google_compute_backend_service.playground_sbs.id
  }

  path_matcher {
    name = "bpaths"
    default_service = google_compute_backend_service.playground_bbs.id
  }
```

### Debug Output

```
- host_rule {
          - hosts        = [
              - "sdwr.domain.tld",
            ] -> null
          - path_matcher = "spaths" -> null
   }
+ host_rule {
          + hosts        = [
              + "beta.domain.tld",
            ]
          + path_matcher = "bpaths"
    }
 + host_rule {
          + hosts        = [
              + "sdwr.domain.tld",
            ]
          + path_matcher = "spaths"
    }
```

### Expected Behavior

```
host_rule {
        hosts        = [
            "sdwr.domain.tld",
        ]
        path_matcher = "spaths"
}
+ host_rule {
          + hosts        = [
              + "beta.domain.tld",
            ]
          + path_matcher = "bpaths"
    }
```

### Additional Context

TL;DR
Every part of new host_rule i add, the terraform will re-sort the host_rule order. It will make the terraform looks like removing existing host_rule and then add back the existing host_rule and the new host_rule in the same time.
I am affraid to use this in production, because i have 39 host_rule.
I cannot imagine if i add the 40th host_rule and terraform shows it removes 39 host_rule and then add 39 host_rule back + 1 new host_rule...
Magicly, when i apply it, the existing hosts aren't removed. So, it's like the terraform removes them only in terraform side but not in google cloud side
I have opened it in [Google Cloud Provider](https://github.com/hashicorp/terraform-provider-google/issues/7859)
host_rule uses [TypeSet](https://github.com/hashicorp/terraform-provider-google/blob/ee251c46904022e68f857408f2192295d3f18ecf/google/resource_compute_url_map.go#L687) and i think The [TypeSet behaviour](https://www.terraform.io/docs/extend/schemas/schema-types.html#typeset) is on terraform site not provider site.

**Comment by jbardin** [comment]
Hi @merceskoba 

Yes, this can be an unfortunate behavior with this schema combination. In most cases terraform does not show changed elements within a set when constructing the diff output. What is happening here is that there must be other "unimportant" change which are not shown here, causing these set elements to change their value. While we have some heuristics to elide some lines of these trivial changes, it still causes the entire set value to change  which needs to be displayed. The new output in 0.14 should at least present the change sin a sorted order to make them easier to spot.

The real culprit here is the legacy provider SDK, which cannot handle the full rich type system that is used within terraform. The impedance mismatch often leaks some changes, most often `null` values being replaced with empty values of the same type, and we work around those to the best of our abilities. There is currently work being done to develop a new SDK, which will have direct access to the same type system, giving providers more control of the data structures they store in the state.

Since this is primarily caused by the plugin SDK (though providers can sometimes work around it in their side), and there's not much more we can do in terraform core until a new SDK is in use, I'm going to close this out for now.

Thanks!
